roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4307533&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4307533&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on master @ [8548987813ff9e1b8a9878023d3abfc6911c16db](https://github.com/cockroachdb/cockroach/commits/8548987813ff9e1b8a9878023d3abfc6911c16db):


```
														ORDER BY
															tab_9072.crdb_internal_mvcc_timestamp
														LIMIT
															1:::INT8
													)
												)
													AS col_15071
											FROM
												defaultdb.public.table1@table1_expr_key AS tab_9070
											WHERE
												'\xe29883':::BYTES::BYTES
												NOT IN (
														SELECT
															'\xbb29ed960621':::BYTES::BYTES AS col_15069
														FROM
															defaultdb.public.table1@table1_expr_key AS tab_9071
														WHERE
															false
														LIMIT
															84:::INT8
													)
											LIMIT
												1:::INT8
										),
										ARRAY[]:::RECORD[]
									)
								),
								(ARRAY[(),(),(),()]),
								(ARRAY[(),(),()]),
								(ARRAY[(),(),(),(),(),(),()])
						)
							AS tab_9074 (col_15072)
				)
		SELECT
			tab_9075.tableoid AS col_15073,
			'{"bar": {"0ih``": {"b": 1.1607717734160394}}, "foo": [[{":DE b5I": {}, "qc}7EGv3": {"bar": [{}], "baz": null}, "u[[": []}, {}], {"foo": 0.07430659923456653, "foobar": {}}]}':::JSONB
				AS col_15074,
			1727859127:::OID AS col_15075,
			'[null, null]':::JSONB AS col_15076,
			ARRAY[]:::RECORD[] AS col_15077,
			261803.805151728437:::DECIMAL AS col_15078,
			1.886734687860280106E+35:::DECIMAL AS col_15079
		FROM
			with_3656 AS cte_ref_1089,
			defaultdb.public.table1@table1_expr_key AS tab_9075,
			defaultdb.public.table1@[0] AS tab_9076
		WHERE
			true
		LIMIT
			65:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4311365&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4311365&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on master @ [8548987813ff9e1b8a9878023d3abfc6911c16db](https://github.com/cockroachdb/cockroach/commits/8548987813ff9e1b8a9878023d3abfc6911c16db):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:258,sqlsmith.go:297,test_runner.go:779: error: pq: internal error: runtime error: index out of range [0] with length 0
		stmt:
		SELECT
			tab_12855.col1_7 AS col_29705
		FROM
			defaultdb.public.table1@table1_col1_2_key AS tab_12855
		GROUP BY
			tab_12855.col1_7
		LIMIT
			62:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/xform/limit_funcs.go
+++ b/pkg/sql/opt/xform/limit_funcs.go
@@ -233,6 +233,12 @@ func (c *CustomFuncs) GenerateLimitedTopKScans(
 	grp memo.RelExpr, sp *memo.ScanPrivate, tp *memo.TopKPrivate,
 ) {
 	required := tp.Ordering
+	// If the ordering was already optimized out (e.g., there is only one possible
+	// value for what would have been the required ordering), then there is no
+	// benefit to exploring limited top K.
+	if len(required.Columns) == 0 {
+		return
+	}
 	// Iterate over all non-inverted and non-partial secondary indexes.
 	var pkCols opt.ColSet
 	var iter scanIndexIter

--- a/pkg/sql/opt/xform/testdata/rules/limit
+++ b/pkg/sql/opt/xform/testdata/rules/limit
@@ -2093,3 +2093,30 @@ top-k
  ├── ordering: +2
  └── scan defg
       └── columns: d:1 e:2 f:3 g:4
+
+# Regression testing for #76102.
+exec-ddl
+CREATE TABLE tab_76102 (
+ a INT2 NULL,
+ b INT2 NULL AS (a + NULL) STORED,
+ UNIQUE (a DESC)
+)
+----
+
+# The GROUP BY column can only have one value (NULL), so  we do not benefit from
+# limited TopK, and trying to apply the rule may lead to errors since the
+# required ordering is optimized away in this case.
+opt expect-not=GenerateLimitedTopKScans
+SELECT b FROM tab_76102@tab_76102_a_key GROUP BY b LIMIT 5
+----
+index-join tab_76102
+ ├── columns: b:2
+ ├── cardinality: [0 - 1]
+ ├── key: ()
+ ├── fd: ()-->(2)
+ └── scan tab_76102@tab_76102_a_key
+      ├── columns: rowid:3!null
+      ├── limit: 1
+      ├── flags: force-index=tab_76102_a_key
+      ├── key: ()
+      └── fd: ()-->(3)

